### PR TITLE
a8n: Add a prototype campaign type to search & replace leaked NPM credentials.

### PR DESCRIPTION
**Pull request by mrnugget**
This is a rough prototype that implements the first steps of #6852. I built this to illustrate how it can work with the current architecture and, more importantly, what we could change to make it more performant and robust.

**Note to reviewers:** Look at this by commits. I had to create one huge commit first that exposes types from the `graphqlbackend` so I can access search

Shortcomings of this PR:

* The campaign type has to _search again_, even though we already searched for the tokens before constructing the `CampaignJob`s. **Possible solution**: persist the `FileMatches` along with `LineMatches` on the `CampaignJob` in a `jsonb` column.
* The diff is created by hand and thus (judging by past "diff does not apply" experiences) very brittle. **Possible solutions**: shell out to `diff`/`git diff` once we have "old file content" and "new file content" at hand, or use a library that constructs a diff in memory (so far, none of them were able to produce a working unified diff), or build a library that does this...
* The search queries are constructed by concatenating strings. That _can_ be brittle, since we concatenate the `scopeQuery`, a user-defined string, to our search terms. We're also missing "count:9999" here to make sure we get everything. **Possible solutions**: something along the lines of [RFC 75](https://docs.google.com/document/d/1iAWIOalYTZLRWmYw0r6_r_uThutLhanyTrxFSFtQAgE/edit) that proposes a more structured interface to search.

We now need to decide which one of these shortcomings is a blocker.

---

Update [in this comment](https://github.com/sourcegraph/sourcegraph/pull/6893#issuecomment-559687549)

### CODE REVIEW

File: enterprise/internal/a8n/service.go [hunk 1]
```diff
@@ -168,7 +168,9 @@ func (s *Service) runChangesetJob(
 		// We use unified diffs, not git diffs, which means they're missing the
 		// `a/` and `/b` filename prefixes. `-p0` tells `git apply` to not
 		// expect and strip prefixes.
-		GitApplyArgs: []string{"-p0"},
+		// Since we also produce diffs manually, we might not have context lines,
+		// so we need to disable that check with `--unidiff-zero`.
+		GitApplyArgs: []string{"-p0", "--unidiff-zero"},
```
Comment (rvantonder) [inline enterprise/internal/a8n/service.go]: What does the patch look like that requires the `--unidiff-zero` bit? I read the description about what the flag does; it's unclear whether we should be addressing the underlying issue, or if this is a `git` thing. LMK if I can look into something.

Comment (rvantonder) [reply enterprise/internal/a8n/service.go]: > Since we also produce diffs manually

Is it only for the manual creation? (I'm not 100% sure what the 'manual' operation is here.)

Comment (eseliger) [reply enterprise/internal/a8n/service.go]: as far as I understood it, the problem is that the generated diff doesn't include surrounding (context) lines, which is not allowed by default

ie

```
this should be
removed
now
```

```diff
- removed
+ added
```

vs

```diff
this should be
- removed
+ added
now
```

Comment (eseliger) [reply enterprise/internal/a8n/service.go]: manual means not diffs that are generated by git, but using string building afaik

File: cmd/frontend/graphqlbackend/search.go [hunk 2]
```diff
@@ -124,6 +124,10 @@ func (r *schemaResolver) Search(args *searchArgs) (searchImplementer, error) {
 	}, nil
 }
 
+func (r *schemaResolver) Search(args *SearchArgs) (SearchImplementer, error) {
+	return NewSearchImplementer(args)
+}
+
```
Comment (rvantonder) [inline cmd/frontend/graphqlbackend/search.go]: To help me understand: What is the rationale behind this change (the wrapping it in `New` part and rename)?

**Comment by rvantonder** [comment]
> I built this to illustrate how it can work with the current architecture and, more importantly, what we could change to make it more performant and robust.

Is the intent to actually merge this PR, or is it just for comments, or possibly both? Would just help me to understand how to approach this PR.

File: enterprise/internal/a8n/service.go (hunk 1 above)
Comment (rvantonder) [reply enterprise/internal/a8n/service.go]: OK, so sort of as if I took a unified diff, and then stripped the context lines, context header, and paths? If that's the case, yes, this will be very brittle, and also potentially ambiguous, without knowing the "before" -> "after" truth and generating a diff from that.

File: web/src/enterprise/campaigns/detail/CampaignDetails.tsx [hunk 3]
```diff
@@ -89,7 +89,7 @@ export const CampaignDetails: React.FunctionComponent<Props> = ({
     // State for the form in editing mode
     const [name, setName] = useState<string>('')
     const [description, setDescription] = useState<string>('')
-    const [type, setType] = useState<'manual' | 'comby'>('manual')
+    const [type, setType] = useState<'manual' | 'comby' | 'credentials'>('manual')
```
Comment (rvantonder) [inline web/src/enterprise/campaigns/detail/CampaignDetails.tsx]: Not just for this line, but generally in this file: As we branch out different campaign types, it would be nice to have a dedicate file of `CampaignTypes` or the like. Inlining the cases everywhere will likely make it hard to keep track of things or reduce readability.

Comment (eseliger) [reply web/src/enterprise/campaigns/detail/CampaignDetails.tsx]: I think this is more supposed to be a prototype rather than anything else, I'm not even sure it was indended to be merged in that state, appreciate your feedback on web code also though :) 
And I agree, if we want to go ahead with merging it i'll refactor that part.

File: enterprise/internal/a8n/campaign_type.go [hunk 4]
```diff
@@ -150,3 +180,76 @@ func (c *comby) generateDiff(ctx context.Context, repo api.RepoName, commit api.
...
+	replacerURL string
+	httpClient  httpcli.Doer
+}
+
+func (c *credentials) searchQuery() string {
+	return c.args.ScopeQuery + " " + npmTokenRegexp + " file:.npmrc"
+}
+
+func (c *credentials) generateDiff(ctx context.Context, repo api.RepoName, commit api.CommitID) (string, error) {
+	searchQuery := "repo:" + regexp.QuoteMeta(string(repo)) + " " + npmTokenRegexp + " file:.npmrc"
```
Comment (rvantonder) [inline enterprise/internal/a8n/campaign_type.go]: > That can be brittle,

Comment (rvantonder) [reply enterprise/internal/a8n/campaign_type.go]: Also I worry about the `QuoteMeta`, because by default in the search query bar we escape the `.`, as in:

`repo:^github\.com/sourcegraph/sourcegraph$ `

And then here, if we take such a string (where users may think the `.` needs to be escaped, or just simply copy paste from the search bar), then:

```
regexp.QuoteMeta(`github\.com/foo/bar`)
```
becomes
`github\\\.com/foo/bar`

Yikes.

Comment (rvantonder) [reply enterprise/internal/a8n/campaign_type.go]: Proposed solution: well, that's difficult. We can choose between being consistent with default search queries (allow regex expressions, but expect escaping. I see this mostly being useful for power users), versus deviating from the former and accept a literal string and sanitize it, so that users do not have to worry about escaping.

I'm personally in favor of deviating from the search query to make this simple, without the regex business, and then reflecting that in the search bar down the line (the `github\.com` has always struck me as visually distracting), and then introducing an alternative syntax for (power) users to indicate a regex intent. This is not at all easy to solve, we'll have to be deliberate and brave about such a decision.

File: enterprise/internal/a8n/service.go (hunk 1 above)
Comment (mrnugget) [reply enterprise/internal/a8n/service.go]: Take a look at the PR description. This is a known problem. I want to explore more options to get better diffs in the next few days, since that would enable us to easily do other things like static-search-&-replace.

Btw. it's only `git apply` that complains. `patch` happily takes the patch :)

Comment (mrnugget) [reply enterprise/internal/a8n/service.go]: > OK, so sort of as if I took a unified diff, and then stripped the context lines, context header, and paths?

No, the context header and paths are still there. The diff produced right now looks like this:

```patch
diff .npmrc .npmrc
--- .npmrc
+++ .npmrc
@@ -1 +1 @@
-//registry.npmjs.org/:_authToken=${NPM_TOKEN}
+//registry.npmjs.org/:_authToken=REMOVED_TOKEN
@@ -2 +2 @@
-//npm.fontawesome.com/:_authToken=12345678-2323-1111-1111-12345670B312
+//npm.fontawesome.com/:_authToken=REMOVED_TOKEN
```

What `git apply` wants, for example, would be this:

```patch
diff .npmrc .npmrc
--- .npmrc
+++ .npmrc
@@ -1,2 +1,2 @@
-//registry.npmjs.org/:_authToken=${NPM_TOKEN}
-//npm.fontawesome.com/:_authToken=12345678-2323-1111-1111-12345670B312
+//registry.npmjs.org/:_authToken=REMOVED_TOKEN
+//npm.fontawesome.com/:_authToken=REMOVED_TOKEN
```

And I'm wondering: since we manually apply the patch to a known revision and we only replace single lines, is it really brittle?

File: cmd/frontend/graphqlbackend/search.go (hunk 2 above)
Comment (mrnugget) [reply cmd/frontend/graphqlbackend/search.go]: I need to access the search implementer from outside the `graphqlbackend` package here: https://github.com/sourcegraph/sourcegraph/pull/6893/files#diff-4b133afba815ca322ac7e402d210dd74R208

In order to do that, I could've either exposed the `schemaResolver` (which I don't need) or expose a way to get to the `SearchImplementer`. So I choose the latter option. Does that make sense? Or did I misunderstand the question?

File: enterprise/internal/a8n/campaign_type.go (hunk 4 above)
Comment (mrnugget) [reply enterprise/internal/a8n/campaign_type.go]: > And then here, if we take such a string (where users may think the `.` needs to be escaped, or just simply copy paste from the search bar), then:

Hmm, but in this case there are no user queries involved. This is solely us searching for the token again. The `QuoteMeta` will only be applied to the repository name that it's in the database.

The only user-defined string involved is the `scopeQuery` to which we add the regex to reduce the number of repositories we have to search again. We can totally avoid that and just search every repository again. That's certainly an option, but I think it's inefficient.

File: cmd/frontend/graphqlbackend/search.go (hunk 2 above)
Comment (rvantonder) [reply cmd/frontend/graphqlbackend/search.go]: OH I see,  to get rid of the receiver. OK 👍

File: enterprise/internal/a8n/campaign_type.go (hunk 4 above)
Comment (rvantonder) [reply enterprise/internal/a8n/campaign_type.go]: > Hmm, but in this case there are no user queries involved. This is solely us searching for the token again



I see, missed that. I thought the `scopeQuery` part might get quoted here.

File: enterprise/internal/a8n/service.go (hunk 1 above)
Comment (rvantonder) [reply enterprise/internal/a8n/service.go]: I see. If it's just single lines this seems fine? Though that `No newline at end of file` issue still applies for the single line manual construction I think? `patch` would at least give a warning, if not fail (I'm not sure which). For the general application of static search and replace, maybe it's worth creating a fallback option that shells out to diff in a temp workspace/file change service, and then we can optimize for in-memory or manually constructing patches as we go along. For _just this_ application, manual construction can probably be made robust enough.

Comment (mrnugget) [reply enterprise/internal/a8n/service.go]: > For the general application of static search and replace, maybe it's worth creating a fallback option that shells out to diff in a temp workspace/file change service, and then we can optimize for in-memory or manually constructing patches as we go along.

That's exactly what I'm going to hack on 😄

**Comment by mrnugget** [comment]
Update on this: if @eseliger is fine with this, I would say that we go ahead and merge this and then refine and improve it in subsequent steps instead of the branch becoming stale.

In the next few days I'm going to look into points (1) and (2) from the original PR description:

1. The campaign type has to search again. Solution I want to investigate: persist the FileMatches along with LineMatches on the CampaignJob in a jsonb column.
2. The diff is created by hand. Solution I want to investigate: shell out to diff/git diff once we have "old file content" and "new file content" at hand.

Besides that, we can also switch to JSON Schema based validation in a follow-up PR, for this campaign type and comby.

File: enterprise/internal/a8n/campaign_type.go [hunk 5]
```diff
@@ -46,26 +42,58 @@ func NewCampaignType(campaignTypeName, args string, cf *httpcli.Factory) (Campai
...
-		return nil, errors.New("missing argument in specification: scopeQuery")
-	}
+		if err := jsonc.Unmarshal(args, &c.args); err != nil {
+			return nil, err
+		}
 
-	if ct.args.MatchTemplate == "" {
-		return nil, errors.New("missing argument in specification: matchTemplate")
-	}
+		if c.args.ScopeQuery == "" {
```
Comment (eseliger) [inline enterprise/internal/a8n/campaign_type.go]: would you want to simplify this logic using the now available json schemas?

Comment (eseliger) [reply enterprise/internal/a8n/campaign_type.go]: nvm didn't see that line on your comment above

**Comment by felixfbecker** [comment]
Please let me review too :)

File: schema/tsconfig.json [hunk 6]
```diff
@@ -5,5 +5,5 @@
     "rootDir": ".",
     "outDir": "out",
   },
-  "include": ["*.json"],
+  "include": ["*.json", "campaign-types/*.json"],
```
Comment (felixfbecker) [inline schema/tsconfig.json]:
```suggestion
  "include": ["**/*.json"],
```

File: web/src/enterprise/campaigns/detail/CampaignDetails.tsx [hunk 7]
```diff
@@ -514,6 +501,16 @@ export const CampaignDetails: React.FunctionComponent<Props> = ({
                             </span>
                         </div>
                     )}
+                    {type && status.state === 'COMPLETED' && (
+                        <div className="d-flex my-3 e2e-preview-success">
+                            <CheckCircleIcon className="icon-inline text-success mr-1" /> Creation complete
```
Comment (felixfbecker) [inline web/src/enterprise/campaigns/detail/CampaignDetails.tsx]: Shouldn't this be "Preview complete"? I looked at the screenshot in Percy and thought it was confusing it says creation complete before having clicked create, when you only wanted to preview

Comment (eseliger) [reply web/src/enterprise/campaigns/detail/CampaignDetails.tsx]: hmm what I meant was "preview creation complete". As this caused confusion, I will name it "Preview complete" and "Creation complete" for either preview or creation

File: schema/tsconfig.json (hunk 6 above)
Comment (eseliger) [reply schema/tsconfig.json]: then it would also require to omit the `out` dir using exclude

```
error TS5055: Cannot write file '/Users/erik/Code/sourcegraph/sourcegraph/schema/out/aws_codecommit.schema.json' because it would overwrite input file.
```

for all .json files, it seems they aren't in the default exclude 🤔 

I thought making it a bit more explicit was easier to understand over omitting `out`

File: web/src/enterprise/campaigns/detail/CampaignDetails.tsx (hunk 7 above)
Comment (felixfbecker) [reply web/src/enterprise/campaigns/detail/CampaignDetails.tsx]: The fact that a preview is "created" is technically correct, but I think from the user pov we want to make it feel like a preview is not persistent, i.e. nothing is "created", only "previewed"

File: schema/tsconfig.json (hunk 6 above)
Comment (felixfbecker) [reply schema/tsconfig.json]: `exclude: out` sounds reasonable to me. Think of the next case, someone sees the `campaign-types` folder and thinks "I should put external service schemas also into a folder", and then has to spend time debugging why it doesn't work vs having it "just work"
